### ISSUE TITLE
Node Stack Overflow Error with SCTransform

### ISSUE THREAD

**Issue by vigneshshanmug** [satijalab/seurat#3904]
Dear Seurat Team,

I am attempting to run the spatial vignette for Slideseq V2 but I keep getting a stack overflow error when I run the SCTransform function. I was able to successfully run the function randomly a couple of times but could not replicate it ever since. I am using a standard macbook pro with 16GB memory. I tried to increase the memory allocation using Sys.setenv('R_MAX_VSIZE'=100000000000) but that didn't help. Can you please help? 

```r
devtools::install_github('satijalab/seurat-data')
library(SeuratData)

InstallData("ssHippo")
slide.seq <- LoadData("ssHippo")

plot1 <- VlnPlot(slide.seq, features = "nCount_Spatial", pt.size = 0, log = TRUE) + NoLegend()
slide.seq$log_nCount_Spatial <- log(slide.seq$nCount_Spatial)
plot2 <- SpatialFeaturePlot(slide.seq, features = "log_nCount_Spatial") + theme(legend.position = "right")
wrap_plots(plot1, plot2)

slide.seq <- SCTransform(slide.seq, assay = "Spatial", ncells = 3000, verbose = FALSE)

Error: node stack overflow
Error during wrapup: node stack overflow
Error: no more error handlers available (recursive errors?); invoking 'abort' restart

sessionInfo()
R version 4.0.3 (2020-10-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Rfast2_0.0.8      Rcpp_1.0.5        dplyr_1.0.2       patchwork_1.1.1  
[5] ggplot2_3.3.3     Seurat_3.9.9.9024

loaded via a namespace (and not attached):
  [1] Rtsne_0.15            colorspace_2.0-0      deldir_0.2-3         
  [4] ellipsis_0.3.1        ggridges_0.5.3        rprojroot_2.0.2      
  [7] fs_1.5.0              rstudioapi_0.13       spatstat.data_1.7-0  
 [10] farver_2.0.3          leiden_0.3.6          listenv_0.8.0        
 [13] remotes_2.2.0         ggrepel_0.9.0         fansi_0.4.1          
 [16] codetools_0.2-16      splines_4.0.3         polyclip_1.10-0      
 [19] pkgload_1.1.0         jsonlite_1.7.2        ica_1.0-2            
 [22] cluster_2.1.0         png_0.1-7             uwot_0.1.10          
 [25] shiny_1.5.0           sctransform_0.3.2     compiler_4.0.3       
 [28] httr_1.4.2            RcppZiggurat_0.1.6    assertthat_0.2.1     
 [31] Matrix_1.2-18         fastmap_1.0.1         lazyeval_0.2.2       
 [34] cli_2.2.0             later_1.1.0.1         htmltools_0.5.0      
 [37] prettyunits_1.1.1     tools_4.0.3           rsvd_1.0.3           
 [40] igraph_1.2.6          gtable_0.3.0          glue_1.4.2           
 [43] RANN_2.6.1            reshape2_1.4.4        spatstat_1.64-1      
 [46] scattermore_0.7       vctrs_0.3.6           nlme_3.1-149         
 [49] lmtest_0.9-38         stringr_1.4.0         globals_0.14.0       
 [52] ps_1.5.0              testthat_3.0.1        mime_0.9             
 [55] miniUI_0.1.1.1        lifecycle_0.2.0       irlba_2.3.3          
 [58] devtools_2.3.2        goftest_1.2-2         future_1.21.0        
 [61] MASS_7.3-53           zoo_1.8-8             scales_1.1.1         
 [64] promises_1.1.1        spatstat.utils_1.20-2 parallel_4.0.3       
 [67] RColorBrewer_1.1-2    curl_4.3              memoise_1.1.0        
 [70] reticulate_1.18       pbapply_1.4-3         gridExtra_2.3        
 [73] rpart_4.1-15          stringi_1.5.3         desc_1.2.0           
 [76] pkgbuild_1.2.0        rlang_0.4.10          pkgconfig_2.0.3      
 [79] matrixStats_0.57.0    lattice_0.20-41       ROCR_1.0-11          
 [82] purrr_0.3.4           tensor_1.5            labeling_0.4.2       
 [85] htmlwidgets_1.5.3     Rfast_2.0.1           cowplot_1.1.1        
 [88] tidyselect_1.1.0      processx_3.4.5        parallelly_1.23.0    
 [91] RcppAnnoy_0.0.18      plyr_1.8.6            magrittr_2.0.1       
 [94] R6_2.5.0              generics_0.1.0        pillar_1.4.7         
 [97] withr_2.3.0           mgcv_1.8-33           fitdistrplus_1.1-3   
[100] survival_3.2-7        abind_1.4-5           tibble_3.0.4         
[103] future.apply_1.7.0    crayon_1.3.4          KernSmooth_2.23-17   
[106] plotly_4.9.2.2        usethis_2.0.0         grid_4.0.3           
[109] data.table_1.13.6     callr_3.5.1           digest_0.6.27        
[112] xtable_1.8-4          tidyr_1.1.2           httpuv_1.5.4         
[115] munsell_0.5.0         viridisLite_0.3.0     sessioninfo_1.1.1  
```

**Comment by saketkc** [comment]
Hi, can you paste the entire `traceback()`? If it is memmory issue, you can try setting `conserve.memory=TRUE` inside `SCTransform`.

**Comment by vigneshshanmug** [comment]
Thanks for the reply. Here is the traceback which is not very helpful. Setting conserve.memory=TRUE does not resolve the problem. 

```
traceback()
No traceback available 
```

**Comment by saketkc** [comment]
Thanks. What's your `options()$expression`? and does this also happen with `ncells=1000`?

**Comment by vigneshshanmug** [comment]
options()$expression was initially 5000 but the error persists after increasing it. It also happens with ncells=1000

```
options()$expression
[1] 5000
> options( expressions = 5e5 )
> options()$expression
[1] 500000
> slide.seq <- SCTransform(slide.seq, assay = "Spatial", ncells = 3000, verbose = FALSE, conserve.memory = TRUE)
Error: node stack overflow
Error during wrapup: node stack overflow
Error: no more error handlers available (recursive errors?); invoking 'abort' restart
> slide.seq <- SCTransform(slide.seq, assay = "Spatial", ncells = 1000, verbose = FALSE, conserve.memory = TRUE)
Error: node stack overflow
Error during wrapup: node stack overflow
Error: no more error handlers available (recursive errors?); invoking 'abort' restart
```

**Comment by saketkc** [comment]
I see. It would probably also help to set `verbose=TRUE` to know what step inside SCT is failing. I would also recommend trying this in a fresh R session.

**Comment by vigneshshanmug** [comment]
It seems like it is failing pretty early on as you can see below. I tried running in fresh R sessions multiple times, reinstalling R and R studio and older versions of R and Seurat with no success.

```
slide.seq <- SCTransform(slide.seq, assay = "Spatial", ncells = 1000, verbose = TRUE, conserve.memory = TRUE)
Calculating cell attributes from input UMI matrix: log_umi
Variance stabilizing transformation of count matrix of size 19375 by 53173
Model formula is y ~ log_umi
Get Negative Binomial regression parameters per gene
Using 2000 genes, 1000 cells
  |                                                                        |   0%Error: node stack overflow
Error during wrapup: node stack overflow
Error: no more error handlers available (recursive errors?); invoking 'abort' restart

```

**Comment by saketkc** [comment]
Not sure of the exact origin of the error. But given there is no traceback, it might have to do with your R/Rstudio installation. Does this also happen if you run this in a vanilla R console (not Rstudio).

**Comment by vigneshshanmug** [comment]
I am getting the same error in a vanilla R console. Now I am getting a traceback which is very long so I cannot copy the entire output. Here is the last part of it. I am not sure how I can share the entire traceback with you. 

```
   "GGGGCTCCCTTAGT", "GGTTTTGTACCGAT", "AAGCAGCGGCCAAT", "ACTCTGACGACATA",
   "CGTTGGCGGACAGT", "CGACCAATCGTCGC", "AGCGAAGTTTTCGA", "GACTCGATAGACGC",
   "TGGTTGGACGCAAG", "CTATCGGATTACCT", "CTTCCCCACCAAAA", "TATTGGGGTATTTA",
   "AACCCCCTAATATC", "AACGGTTTGGCCCC", "ATAAGGCTGCTTGA", "CATATTTTTGTACC",
   "CGCGCCCCGGGGCC", "CGCTTGGCGATTTC", "ATCTCAAGCATAAA", "GGTGCAAAAACTAA",
   "GCATTGGCTTAATG", "CAGAATATTGAGGG", "GACTTAACGGACCC", "GGTGGGGGTTTTTG",
   "ACCAATTACAGATA", "AAGTTAGGCTACAA", "TCAACCAACGGAGA", "TATCATACTGTTTT",
   "AAATTGATGGGCCA", "AACGGTAATACGTG", "GATAGAAATGATGC", "TGCCCTTACCAAAA",
   "ACTTCTAGGAATGA", "CTCCCACCTATGAC", "GGAACTCGACTCAG", "GGGTCCACCCGCGG",
   "TTCGTCCCCGTCTA", "CAATATTTAAGTCA", "TGTGACTGCGGAGG", "AGTCGGTCCAGTAC",
   "CATGTGGGGAGTGA", "GGTAGTTGGCAGAC", "CCCGCCAAATCACG", "GACATCTATGAAAA",
   "TAACTCGGCGGAAG", "ATATATTCGCAGTT", "CTTTCCGATCGGAA", "AAACGCCCGAGAAA",
   "GTGACGAACCAAAA", "AAATTGAGATCACC", "GCGCATACATTAGT", "ATAAGGGGTGAGTT",
   "CTGTGCAGGGTTCA", "CTTTGCCCGAGTGT", "AAAATATTAAGCCC", "ACAATACAGATACT",
   "GGTCAGGTTAACGT", "CATTATGGAATTCT", "GACTTTTCTTAAAG", "GTCAATAAAGGGCG"
   ), class = "data.frame"), verbosity = 0, return_cell_attr = TRUE,
       return_gene_attr = TRUE, return_corrected_umi = TRUE, n_cells = 3000)
2: do.call(what = "vst", args = vst.args)
1: SCTransform(slide.seq, assay = "Spatial", ncells = 3000, verbose = FALSE)
```

**Comment by ChristophH** [comment]
Since this looks like an sctransform problem, I'd be happy to have a look. Can you share the data that triggers the error? (either here or in an email to christoph.hafemeister@nyu.edu)

**Comment by vigneshshanmug** [comment]
Hi ChristophH, 

Thank you for looking into this. I am having the issue with the ssHippo data on SeuratData and my own similar datasets. Here is the code I used to download the ssHippo data. 

`devtools::install_github('satijalab/seurat-data')
library(SeuratData)

InstallData("ssHippo")
slide.seq <- LoadData("ssHippo")`

**Comment by ChristophH** [comment]
I cannot reproduce the error on my end
```
slide.seq <- LoadData("ssHippo")
set.seed(42)
slide.seq <- SCTransform(slide.seq, assay = "Spatial", ncells = 1000, verbose = TRUE, conserve.memory = TRUE)
```
runs just fine. I am using `Seurat_3.9.9.9008` and `sctransform_0.3.2.9001` (Seurat release/4.0.0 and sctransform develop branch)

**Comment by vigneshshanmug** [comment]
Thank you all for your help. I tried everything that was suggested but I cannot get it to work on my computer but I was able to get the SCTransform function working on a virtual machine. So I suspect this issue is specific to my computer or version of R/Rstudio.

**Comment by duocang** [comment]
@vigneshshanmug @ChristophH  @saketkc 

Hi.

This is actually very common from what I experienced (Windows).  I meet this problem many times.

The only solution I have is to put all source files and data into a new folder.  Dull but no other options.

I just have the problem again in my Windows (96 GB for only 4000 cells) but when I run the script with VMware in the same folder. NO PROBLEM at all.


I have struggled with this problem for months and no idea at all.

Best regards
Xuesong
